---
title: "Introduction to Python, Session 1 <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_course_template/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
knitr::opts_chunk$set(echo = TRUE, tidy = T)
```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# Introduction to Python

---
"    
  )
  
}

```

```{r}
library(IntroToPython)
library(reticulate)
py_install(c("matplotlib", "pandas"))
```

```{python}

import numpy
import pandas

```

```{python}

s = pandas.Series([1, 3, 5, numpy.nan, 6, 8])
s

```

```{python}
import matplotlib
import matplotlib.pyplot as plt

```

```{python}
fig, ax = plt.subplots()

fruits = ['apple', 'blueberry', 'cherry', 'orange']
counts = [40, 100, 30, 55]
bar_labels = ['red', 'blue', '_red', 'orange']
bar_colors = ['tab:red', 'tab:blue', 'tab:red', 'tab:orange']

ax.bar(fruits, counts, label=bar_labels, color=bar_colors)

ax.set_ylabel('fruit supply')
ax.set_title('Fruit supply by kind and color')
ax.legend(title='Fruit color')

plt.show()
```

## Overview

??We often put an overview slide that links to the various parts of the course content to make it easy to navigate i.e. this one for intro to r??

- [Set up](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#set-up)
- [Background to R](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#background-to-r)
- [Data types in R](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#data_types_in_r)
- [Reading and writing in R](https://rockefelleruniversity.github.io/Intro_To_R_1Day/r_course/presentations/singlepage/introToR_Session1.html#reading-and-writing-data-in-r)

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Set Up

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Set Up

---
"    
  )
  
}

```

## Materials

??We try to denote each major section of content with these title slides (see above for set up). The hierarchy will help with the different versions of the teaching content i.e. slides and single page.??

?? Materials just gives the links to download the package Just ensure links are up to date??

All prerequisites, links to material and slides for this course can be found on github.

* [Intro_To_R_1](https://rockefelleruniversity.github.io/Intro_To_R_1Day/)

Or can be downloaded as a zip archive from here.

* [Download zip](https://github.com/rockefelleruniversity/Intro_To_R_1Day/zipball/master)

---
## Course materials

?? this slide just shows where thye can get the course content from the download. You do not need to populate these directories. It will be made during compilation??

Once the zip file in unarchived. All presentations as HTML slides and pages, their R code and HTML practical sheets will be available in the directories underneath.

* **r_course/presentations/slides/**
Presentations as an HTML slide show.
* **r_course/presentations/singlepage/** 
Presentations as an HTML single page.
* **r_course/presentations/r_code/**
R code in presentations.
* **r_course/exercises/**
Practicals as HTML pages. 
* **r_course/answers/**
Practicals with answers as HTML pages and R code solutions. 


---
## Set the Working directory


Before running any of the code in the practicals or slides we need to set the working directory to the folder we unarchived. 

You may navigate to the unarchived RU_Course_help folder in the Rstudio menu.

**Session -> Set Working Directory -> Choose Directory**

or in the console.

```{r setwd_introtoR,eval=F} 
setwd("/PathToMyDownload/RU_Course_template/r_course")
# e.g. setwd("~/Downloads/Intro_To_R_1Day/r_course")
```


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Background of Python

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Background of Python
      
---
"    
  )
  
}

```

## What is Python?

Python is a high-level, general-purpose programming language.

Guido van Rossum began working on Python in the late 1980s as a successor to the ABC programming language and first released it in 1991 

One of the most noticeable differences comes from the emphasis on code readability:
it uses significant indentation 




Python 3.0, released in 2008, was a major revision not completely backward-compatible with earlier versions.While Python 2.7 and older is officially unsupported, a different unofficial Python implementation, PyPy, continues to support Python 2, i.e

o index

---
## What is Python to you?

.pull-left[
Python has a huge user base across many disciplines. These users have developed a wide range of libraries which you can use.  

As python is used by many fields, it is a useful language for adapting novel approaches for bioinformatics:
such as deep learning methods for scRNAseq or Image Analysis
  ]
  
.pull-right[

https://en.wikipedia.org/wiki/TIOBE_index
  ]

---
## Pyhton vs R

Though R has long been the key coding language in bioinformatics, python is growing in use. 

Though there are core utility packages that have been around a long time such as Biopython. It is the new technique-specific packages that are driving the surging popularity of python i.e scanpY 

---
## Pyhton vs R

Table here of strangths of each


https://microbenotes.com/python-r-programming-language-bioinformatics/
https://medium.com/@hamza_33678/programming-for-bioinformatics-r-vs-python-52969a1f7a49

 Biopython, ML and scRNAseq
 image analysis


---
## Setting Up Python

ensure installed
interactive vs scripts

---
## Navigating VS Code


---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Data Types and Functions in Python

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Data Types and Functions in Python
      
---
"    
  )
  
}

```



## Simple Calculation

At its simplest you can just use python as a fancy calculator:

```{python}
1+1
```
```{python}
2*5
```

---
## Functions

To take things further there are many functions built-in to python. These are saved chun ks of code that will do a task based on the arguments you provide. You can tell there is a function when there is a string immediately followed by a set of parenthesis:

*myfunction()*

Here we use the round function:

```{python}
round(3.14159)
```

---
## Help with functions

To get help with a function you can use the *help()* function. This will open up the help page for this function. Hopefully this will contain information about what arguments it are accepted and what is returned by the function. In this case we can see there is an additonal optional argument *ndigits*. This has a default value of **None**.

```{python}
help(round)
```

---
## Function and arguments

If we want to update our rounding result to allow for more decimal places we can add the second argument. For simple function like this, s long as the order is correct we do not to specify the argument. 

```{python}
round(3.14159, 3)
```

We can still run a function with disordered arguments by naming them. 

```{python}
round(ndigits=3, number=3.14159)
```

 
---
## Variables

Often you will want to save something in your environment for use later on. We do this by creating a variable by assignment with the **=** sign.

```{python}
greeting = 'Hello!'
greeting
```

```{python}
number = 3.14159
number
```

---
## Variables

When assigning a variable there are certain things you can and cannot do.

Good:

* Only letters, numbers, and the underscore

Bad:

* Don't use non-alphanumerics. This includes: . % + - * #  
* You can't start with a number

The best thing to do is name it something short and simple, that makes sense.

---
## Variables and functions

Once we have a variable we can then use it inside functions. The vector name is acting as an alias for what it contains.

```{python}
number
```

```{python}
round(number)

```


---
## Variables

There are many kinds of variables. The most basic types are: *str*,*float* and *int*. You can always check what kind you have with the *type()* function.

```{python}
greeting = 'Hello!'
type(greeting )
```

```{python}
number = 3.14159
type(number )
```

```{python}
newnumber = round(number)
newnumber
```
```{python}
type(newnumber)
```

---
## Variables Coercion

We can manually set the type using: *str()*, *float()* and *int()*. 

```{python}
string_number = str(number)
string_number
```

```{python}
float_string_number = float(string_number)
float_string_number 

```

```{python}
int_float_string_number = int(float_string_number)
int_float_string_number 

```

---
## Variables Coercion

These functions do not always work, if there is not a clear rationale for how to resolve the function.

```{python, eval=F}
greeting = 'Hello!'
int(greeting)
```

```{python, eval=T, echo=F}

try:
  float(greeting)
except Exception as e:
  print(e)
  
```

---
## Concatenation

Strings can be concatenated easily.

```{python}
newgreeting = 'Hi' ' there'
newgreeting
```
```{python}
newgreeting2 = 'Hi' + ' there'
newgreeting2
```
```{python}
newgreeting3 = 'Hi'
newgreeting3 += ' there'
newgreeting3 
```

```{python}
newgreeting4 = 'Hi' * 5
newgreeting4
```

---
## Lists

Python has many options for storing data. The simplest is a list. Lists are denoted with square brackets. 

```{python}

my_strs = ['a','b','c','d','e']
my_strs

```

```{python}

my_ints= [1,2,3,4,5]
my_ints

```

```{python}

my_floats = [1.1,2.2,3.3,4.4,5.5]
my_floats

```

---
## Indexing Lists

We can also use the square brackets to extract specific values from our list. 

```{python}
my_strs[2]
```

Here we get the third value from our list using the number 2. That is because python uses zero indexing; counting in python starts at 0, not 1. 

```{python}
my_strs[0]
```

---
## Indexing Lists

Sometimes we have a long list, but we know we want the final value. We can use a *-* to indicate how far from the end we want to index.

```{python}
my_strs[-1]
```

We can also create a sublist by slicing our list with the *:*.

```{python}
my_strs[2:4]
```

```{python}
my_strs[2:-1]
```

```{python}
my_strs[2:]
```


```{python}
my_strs[0] = 'z'
```

---
## Concatenation

We can concatenate lists, just as we did with strings.

```{python}
biglist = my_strs + my_ints
biglist
```

```{python}
biglist = my_strs 
biglist += my_ints
biglist
```

```{python}
biglist = my_strs * 5
biglist
```

---
## Functions and lists

There are many useful functions for working with lists. Many of these functions work directly on the list. This means you don't need to assign the result back to the object. The structure is VARIABLE.function(). Here we use the append function to extend our list.

```{python}
my_strs.append('f')
my_strs
```

```{python}
my_strs.append(1)
my_strs
```

---
## Functions and lists

There are many useful functions built into the base version of python. 



* .insert() inserts an argument into a specfic postion in the list

```{python}
my_strs.insert(3,'c')
my_strs
```

* remove() inserts an argument into a specfic postion in the list

```{python}
my_strs.remove('c')
my_strs
```

* del() inserts an argument into a specfic postion in the list

```{python}
del(my_strs[3])
my_strs
```


* .index() reveals which position in the list is the supplied argument

```{python}
my_strs.index('c')
```



<!-- .sort() -->
<!-- sorted(furniture) -->

<!-- furniture = ['table', 'chair', 'rack', 'shelf'] -->
<!-- table, chair, rack, shelf = furniture -->
<!-- unique? -->

<!-- Remember! -->
<!-- Square brackets [] for indexing and lists. -->

<!-- x[1] -->
<!-- ## [1] 1 -->
<!-- Parentheses () for function argments. -->

<!-- sqrt(4) -->
<!-- ## [1] 2 -->

<!-- logical operations -->

---
## Tuples



The key difference between tuples and lists is that, while tuples are immutable objects, lists are mutable. This means that tuples cannot be changed while the lists can be modified. Tuples are more memory efficient than the lists.

---
## Variables


    - range
    - indexing
    - bool: logicals
  * tuple
  * list
  * dictionary
- reading/writing


- libraries
- control and list comp










## Slide Structure

We use the a double has/pound for slides. This is important for hierarchy. If you use 1 it will work for the slides, but will mess up the single page version.


## Left/Right alignment

To show things side by side on a slide you can use pull left/right. t is important to keep the tab on the second square bracket. On single page it will just be one after the other:

.pull-left[
This will be on left of slide
  ] 
  
.pull-right[
This will be on right

 ]


## Data from external sources

everything in the rmd is relative to the inst/extdata directory so you have some data to load in, just put it in the data directory, read it in along the path. 
The structure is maintained for those attending the course and download the package using the instructions. The same rules are followed for imgs and scripts that you may call.

```{py}
import 

```

--
## Exercises
The following few slides show you how to structure exercise slides.

We often have several exercise slides per session. So you can just copy and paste and change the directory to the appropriate name. All 3 file types are made from you single exercise Rmd. 

---
## Time for an exercise!

??? exercise description here?? [here](../../exercises/exercises/MyExercise1_exercises.html)

---
## Answers to exercise

Answers can be found [here](../../exercises/answers/MyExercise1_answers.html)

R code for solutions can be found [here](../../exercises/answers/MyExercise1_answers.R)

## Exercise again

So we can have some more sections here describing some new important topic

Then more exercises

---
## Time for an exercise!

??? exercise description here?? [here](../../exercises/exercises/MyExercise2_exercises.html)

---
## Answers to exercise

Answers can be found [here](../../exercises/answers/MyExercise2_answers.html)

R code for solutions can be found [here](../../exercises/answers/MyExercise2_answers.R)
