---
title: "Exercise 5"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/Intro_To_Python/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```


<p>&nbsp;</p>
These exercises are about logical operators and reading/writing and scripts from [Session 2](https://rockefelleruniversity.github.io/Intro_To_Python/). 


**Exercise 1** - **Reading/Writing**

- Read in the tab delimited file "GeneExpression.txt". (Hint: The delimiter for a tab file is "\t".)
```{python,echo=toMessage}
import numpy as np

geneExpression = np.genfromtxt("data/GeneExpression.txt", delimiter="\t", skip_header=True)

```

- Read in the tab delimited file "GeneExpressionWithMethods.txt". This file contains information on analysis steps used to produce file. We will want to skip those. 
*Hint: Check the `skip_header` argument in the help page*

```{python,echo=toMessage}

geneExpression = np.genfromtxt("data/GeneExpressionWithMethods.txt", delimiter="\t", skip_header=4)

geneExpression
```

- Find the mean expression across rows. (Hint: NumPy objects have a mean function attribute)

```{python,echo=toMessage}

geneExpression[0,].mean()
geneExpression[1,].mean()
geneExpression[2,].mean()
geneExpression[3,].mean()
geneExpression[4,].mean()
geneExpression[5,].mean()
geneExpression[6,].mean()
geneExpression[7,].mean()

# Alternatively we could use the axis argument.
geneExpression.mean(axis=1)
```

- The corresponding gene names are stored in "GeneNames.txt". Read this in and subset it based on whether the mean gene expression is over 6. Write this out into a new text file. 

```{python,echo=toMessage}

sub_idx = geneExpression.mean(axis=1) > 6
geneNames = np.genfromtxt("data/GeneNames.txt", delimiter="\t", dtype="U6")
geneNames_sub = geneNames[sub_idx]
np.savetxt("GeneNames_highexpression.txt", geneNames_sub, delimiter="\t", fmt='%s')
```


**Exercise 2** - **Scripts**

Lets try to put as much together that we have learnt thus far. This will be a multistep challenge. Break it down and use pseudocode to help. Start by working the code interactively, then turn it into a script. 


1. Read in the "data/GeneExpressionWithMethods.txt" dataset. 
2. Use a for loop to calculate the Z score for each gene (per row). The zscore is (gene_expression - mean)/standard deviation. You should use a function to do this calculation.
2. Save the result as a NumPy array. The *vstack* function might be useful. 
3. Find which gene has the highest absolute max Zscore. This is a very rough proxy for the variability of that gene.
4. Print out the gene name with the highest value
5. Turn this into a script and run the script


```{python,echo=toMessage}

geneExpression = np.genfromtxt("data/GeneExpressionWithMethods.txt", delimiter="\t", skip_header=4)
```

```{python,echo=toMessage}
my_mean = geneExpression.mean(axis=1)
my_std = geneExpression.std(axis=1)
```

```{python,echo=toMessage}

def zscore(value, mean, std):
  my_z = (value-mean)/std
  return my_z

```

```{python,echo=toMessage}

for i in range(geneExpression.shape[0]):
  geneExpression_zscore = zscore(geneExpression[i], my_mean[i], my_std[i])
  if i==0:
    my_zscore=np.array(geneExpression_zscore)
  else:
    my_zscore=np.vstack((my_zscore,geneExpression_zscore))
  
my_abs = abs(my_zscore)
top_values = my_abs.max(axis=1)
top_value = top_values.max()
my_top_index = top_values == top_value

geneNames = np.genfromtxt("data/GeneNames.txt", delimiter="\t", dtype="U6")
geneNames_sub = geneNames[my_top_index]
print(geneNames_sub)

```
