---
title: "My Exercise 2"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/Intro_To_Python/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```


<p>&nbsp;</p>
These exercises are about data types and custom functions from Session 2 [session1](https://rockefelleruniversity.github.io/Intro_To_Python/). 

**Exercise 1** - **Reading/Writing**

- Read in the tab delimited file "GeneExpression.txt". (Hint: The delimiter for a tab file is "\t".)
```{python,echo=toMessage}
import numpy as np

geneExpression = np.genfromtxt("data/GeneExpression.txt", delimiter="\t", skip_header=True)

```

- Read in the tab delimited file "GeneExpressionWithMethods.txt". This file contains information on analysis steps used to produce file. We will want to skip those. (Hint: Check the `skip_header` argument in the help page)
```{python,echo=toMessage}

geneExpression = np.genfromtxt("data/GeneExpressionWithMethods.txt", delimiter="\t", skip_header=4)

geneExpression
```

- Find the mean expression  across rows. (Hint: NumPy objects have a mean function attribute)
```{python,echo=toMessage}

geneExpression[0,].mean()
geneExpression[1,].mean()
geneExpression[2,].mean()
geneExpression[3,].mean()
geneExpression[4,].mean()
geneExpression[5,].mean()
geneExpression[6,].mean()
geneExpression[7,].mean()

# Alternatively we could use the axis argument.
geneExpression.mean(axis=1)
```

- The corresponding gene names are stored in "GeneNames.txt". Read this in and subset it based on whether the mean gene expression is over 6. Write this out into a new text file. 

```{python,echo=toMessage}

sub_idx = geneExpression.mean(axis=1) > 6
geneNames = np.genfromtxt("data/GeneNames.txt", delimiter="\t", dtype="U6")
geneNames_sub = geneNames[sub_idx]
np.savetxt("GeneNames_highexpression.txt", geneNames_sub, delimiter="\t", fmt='%s')
```


**Exercise 2** - **If Else**

- Build an if/else statement that test if a variable is odd/even. (Hint: The modulus operator may be useful here i.e. x%2 returns the remainder after the value of x is divided by 2). 


```{python,echo=toMessage}
x=1

if x%2==0:
  print("Even")
else:
  print("Odd")

```

- Put this in a function and try out it out on various variables.

```{python,echo=toMessage}

def even_or_odd(x):
  if x%2==0:
    return "Even"
  else:
    return "Odd"

even_or_odd(10)
even_or_odd(3)
even_or_odd(10007)
```

**Exercise 3** - **Loops**

- Calculate the factorial (factorial of 3 = 3 * 2 * 1) of 10 using a loop. (Hint: first try to write some pseudocode to help you build this)

```{ echo=toMessage, eval=F}

# Idea
Each iteration of a loop I want to multiply my factorial by itself:
x=1  result_1 = 1
x=2  result_2 = result_1*x = 2*1
x=3  result_3 = result_2*x = 3*2*1
x=4  result_4 = result_3*x = 4*3*2*1
etc...

# Pseudocode

for 0 to 9:
  if x is 0:
    create an object and set its value to 1
  else:
    once we are past the first data point we can multiply factorialAnswer by the iteration we are on. Then overwrite factorialAnswer with this result. 

```

```{python,echo=toMessage}

for x in range(10):
  if x == 0:
    factorialAnswer = 1
  else:
    factorialAnswer = factorialAnswer * x

factorialAnswer
```

- Adjusting your answer from before, what is the first number that has a factorial greater than 1000.

```{python,echo=toMessage}

factorialAnswer = 0
count = 0

while factorialAnswer <= 1000:
  count = count+1
  if count == 1:
    factorialAnswer = 1
  else:
    factorialAnswer = factorialAnswer * count 


count
factorialAnswer

```


